README
Copyright (c) 2005,2006,2007,2008,2009,2010 Christopher Kleckner
All rights reserved

This file is part of the Rio library for ruby.

Rio is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

Rio is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Rio; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


= Rio - Ruby I/O Facilitator

fa-cil-i-tate:  To make easy or easier 

Rio is a facade for most of the standard ruby classes that deal with
I/O; providing a simple, intuitive, succinct interface to the
functionality provided by IO, File, Dir, Pathname, FileUtils,
Tempfile, StringIO, OpenURI and others. Rio also provides an
application level interface which allows many common I/O idioms to be
expressed succinctly.

===== Suggested Reading

* RIO::Doc::SYNOPSIS
* RIO::Doc::INTRO
* RIO::Doc::HOWTO
* RIO::Doc::EXAMPLES
* RIO::Rio

==== Installation

To install Rio from the distribution directory type
  ruby setup.rb

Rio is also available as a gem. If you have rubygems installed use:
  gem install rio



Project::       http://rubyforge.org/projects/rio/
Documentation:: http://rio4ruby.com/
Bugs::          http://rubyforge.org/tracker/?group_id=821
Blog::		http://rio4ruby.blogspot.com/
Email::         rio4ruby@rubyforge.org




doc/RELEASE_NOTES
=== Rio - Ruby I/O Facilitator

Rio is a facade for most of the standard ruby classes that deal with
I/O; providing a simple, intuitive, succinct interface to the
functionality provided by IO, File, Dir, Pathname, FileUtils,
Tempfile, StringIO, OpenURI and others. Rio also provides an
application level interface which allows many common I/O idioms to be
expressed succinctly.

fa-cil-i-tate:  To make easy or easier.

To create the documentation for Rio run the command
  ruby build_doc.rb
from the distribution directory.

Then point your browser at the 'doc/rdoc' directory.

Suggested Reading
* RIO::Doc::SYNOPSIS
* RIO::Doc::INTRO
* RIO::Doc::HOWTO
* RIO::Rio
=
== New for version 0.4.4
* Usable for 1.8 code being ported to 1.9.
* This release does not address encoding issues required by Ruby 1.9

== New for version 0.4.3
* Minor bug fixes.
* Release before Branch in development for Ruby 1.9.

== New for version 0.4.2
* Fix tests to better deal with Changes to URI library in ruby 1.8.7.
* Other bug fixes.

== New for version 0.4.1
* Bug fixes to better handle file system paths with URI special
  characters.
* Other bug fixes.

= New for version 0.4.0
* Alpha. 


== New for version 0.3.9
* Bug fixes, code cleanup. 
* See Rio#split


== New for version 0.3.8
* Enhanced support for FTP file-systems. All of Rio's most powerful
  idioms are now supported seamlessly on FTP servers.

   # copy a file from or to an FTP server
   rio('ftp://ahost/adir/afile') > rio('localfile') # server -> local file
   rio('ftp://ahost/adir/afile') < rio('localfile') # local file -> server

   # copy an entire directory structure from or to an FTP server
   rio('ftp://ahost/adir') > rio('localdir') # server -> local directory
   rio('ftp://ahost/adir') < rio('localdir') # local directory -> server

  All of Rio's grande selection and filtering options are available for
  files and directories on FTP servers.

   # create a gzipped copy of a web page on an ftp server
   rio('ftp://ftphost/f.html.gz').gzip < rio('http://httphost/f.html')

   # dump the first 10 lines of a gzipped log file on an FTP server to stdout
   rio('ftp://ftphost/logfile.txt').lines(0...10) > ?-

   # iterate through the entries of a directory on an FTP server
   rio('ftp://ftphost/adir').entries { |entrio| ... }

   # get an array of all .rb files on an ftp server
   rb_files = rio('ftp://ftphost/').all.files['*.rb']

* More tests
* Bug fixes

== Version 0.3.7
* Maintenance release

== New for version 0.3.6
* Pipe operator and command pipes

  Pipe multiple commands
   rio('afile') | rio(?-,'acmd') | 'another_cmd' | ?-

  Run the same series of commands, with different input and/or output
   cmdpipe = rio(?-,'acmd') | rio(?-,'another_cmd')
   rio('infile1') | cmdpipe | rio('outfile1')
   rio('infile2') | cmdpipe | rio('outfile2')

   cmdpipe2 = rio(?|,'cmd1','cmd2',rio('outfile'))   # create a cmdpipe Rio
   rio('infile1') | cmdpipe2       # run with input coming from a file
   rio(?-) | cmdpipe2              # same commands with input from stdin


== New for version 0.3.5
* Partial support for MS Windows drive and UNC path specifications
  
  The separator must be '/'. Back-slashes are not supported.

  Drives and UNC paths may be specified as expected:
   rio("D:/adir/afile")
   rio('//ahost/adir/afile')

  This support has introduced a change in the behaviour
  of Rio#path. Rio#path now returns the path on the 
  file-system for Rios that are on the file-system; returns the path
  portion of the URL for FTP, HTTP and other Rios that
  have a path, and return nil for Rios that have no path
   rio('/a/b/c').path                     #==> "/a/b/c"
   rio('b/c').path                        #==> "b/c"
   rio('C:/b/c').path                     #==> "C:/b/c"
   rio('//ahost/a/b').path                #==> "//ahost/a/b"
   rio('file://ahost/a/b').path           #==> "//ahost/a/b"
   rio('file:///a/b').path                #==> "/a/b"
   rio('file://localhost/a/b').path       #==> "/a/b"
   rio('http://ahost/index.html').path    #==> "/index.html"
   rio('stdin:').path                     #==> nil


== New for version 0.3.4
* New Grande Selection parameter.

  A major weakness of Rio's selection methods (lines, files, etc.)
  has always been that it only implemented a logical OR.

   rio('afile').lines(0..10,/Rio/) {...}
  iterates through lines that are in the range 0..10 OR 
  contain 'Rio'.

   rio('adir').files(:executable?,'*.rb') {...}
  iterates through files that are executable OR match '*.rb'

  Selecting files that matched both required using a proc.
   rio('adir').files(proc{ |f| f.executable? and f.fnmatch?('*.rb')}) {...}

  Rio's grande selection methods will now accept an array of conditions
  which must all be matched, in order to be selected. A logical AND.

   rio('adir').files([:executable?,'*.rb']) {...}
  
  The array, of course, need not be the only paramter. 

   rio('adir').files('*.exe',[:executable?,'*.rb']) {...}

  selects .exe files and .rb files that are executable.

* Renamed some of grande rejection methods. 
  (based on a suggestion by Gavin Sinclair)
   nolines => skiplines
   nofiles => skipfiles
  etc.

* New skip() grande method
   rio('afile').skip.lines(/Rio/)       # same as skiplines(/Rio/)
   rio('afile').lines(/Rio/).skip(0..9) # lines with 'Rio', exclude 
                                        # the first ten lines

* Alternative syntaxes for creating Rios that have no path.

   rio(?-)          # create a Rio refering to stdio
   rio(:stdio)      # same thing.
   rio.stdio        # same thing
   RIO.stdio        # ditto
   RIO::Rio.stdio   # once again

* From Pathname added
  * root?
  * mountpoint?
  * realpath
  * cleanpath

* Removed Rio#slurp in favor of Rio#contents.


* Added aliases for the copy operators. (suggested by Dave Burt)
   * copy_to >
   * append_to >>
   * copy_from <
   * append_from <<


* Bug fixes and corrections    
  
Project:: http://rubyforge.org/projects/rio/

lib/rio.rb
#--
# =============================================================================== 
# Copyright (c) 2005,2006,2007,2008,2009,2010 Christopher Kleckner
# All rights reserved
#
# This file is part of the Rio library for ruby.
#
# Rio is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# Rio is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Rio; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
# =============================================================================== 
#++
#
# To create the documentation for Rio run the command
#  ruby build_doc.rb
# from the distribution directory.
#
# Suggested Reading
# * RIO::Doc::SYNOPSIS
# * RIO::Doc::INTRO
# * RIO::Doc::HOWTO
# * RIO::Doc::EXAMPLES
# * RIO::Rio
#


require 'rio/version'
require 'rio/base'
require 'rio/def'
require 'rio/exception'

require 'forwardable'

$trace_states = false

require 'rio/kernel'
require 'rio/constructor'
require 'rio/construct'

require 'rio/const'

module RIO
  class Rio #:doc:
    require 'rio/local'
    include Local
    require 'rio/factory'
    protected

    attr_reader :state

    public

    # See RIO.rio
    def initialize(*args)
      @state = Factory.instance.create_state(*args)
    end

    def initialize_copy(*args)
      #p callstr("initialize_copy",args[0].inspect)
      super
      @state = Factory.instance.clone_state(@state)
    end

    # See RIO.rio
    def self.rio(*args,&block) # :yields: self
      ario = new(*args)
      if block_given?
        old_closeoncopy = ario.closeoncopy?
        begin
          yield ario.nocloseoncopy
        ensure
          ario.reset.closeoncopy(old_closeoncopy)
        end
      end
      ario
    end

    # Returns the string representation of a Rio that is used by Ruby's libraries.
    # For Rios that exist on the file system this is Rio#fspath.
    # For FTP and HTTP Rios, this is the URL.
    #
    #  rio('/a/b/c').to_s                    ==> "/a/b/c"
    #  rio('b/c').to_s                       ==> "b/c"
    #  rio('C:/b/c').to_s                    ==> "C:/b/c"
    #  rio('//ahost/a/b').to_s               ==> "//ahost/a/b"
    #  rio('file://ahost/a/b').to_s          ==> "//ahost/a/b"
    #  rio('file:///a/b').to_s               ==> "/a/b"
    #  rio('file://localhost/a/b').to_s      ==> "/a/b"
    #  rio('http://ahost/index.html').to_s   ==> "http://ahost/index.html"
    #
    def to_s() target.to_s end

    alias :to_str :to_s
    def dup
      #p callstr('dup',self)
      self.class.new(self.to_rl)
    end
    
    def method_missing(sym,*args,&block) #:nodoc:
      #p callstr('method_missing',sym,*args)
      
      result = target.__send__(sym,*args,&block)
      return result unless result.kind_of? State::Base and result.equal? target
      
      self
    end
    
    def inspect()
      cl = self.class.to_s[5..-1]
      st = state.target.class.to_s[5..-1]
      sprintf('#<%s:0x%x:"%s" (%s)>',cl,self.object_id,self.to_url,st)
    end
    

    protected

    def target()  #:nodoc:
      @state.target 
    end

    def callstr(func,*args) #:nodoc:
      self.class.to_s+'.'+func.to_s+'('+args.join(',')+')'
    end

  end # class Rio
end # module RIO
module RIO
  class Rio
    USE_IF = true #:nodoc:
    
    if USE_IF
      include Enumerable
      require 'rio/if'
      include RIO::IF::Grande
      require 'rio/ext/if'
      include RIO::IF::Ext
    end
  end
end

#require 'rio/ext/zipfile.rb'

if $0 == __FILE__
  eval DATA.read, nil, $0, __LINE__+4
end

__END__

puts
puts("Run the tests that came with the distribution")
puts("From the distribution directory use 'test/runtests.rb'")
puts
